lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -404,41 +530,15 @@ void PatchEntryPointMutate::lowerCpsJump(Function *parent, cps::JumpOp *jumpOp,
   SmallVector<Value *> vgprArgs;
   vgprArgs.push_back(vcr);
   vgprArgs.push_back(vsp);
-  splitIntoDwords(layout, builder, remainingArgs, vgprArgs);
-
-  Value *vgprArg = mergeIntoStruct(builder, vgprArgs);
-
-  // Packing SGPR arguments(user data) into vector of dwords.
-  SmallVector<Value *> userData;
-  for (unsigned idx = 0; idx != numUserdata; ++idx)
-    userData.push_back(parent->getArg(idx));
-
-  SmallVector<Value *> userDataDwords;
-  splitIntoDwords(layout, builder, userData, userDataDwords);
-  Value *userDataVec = mergeDwordsIntoVector(builder, userDataDwords);
-
-  // Mask off metadata bits and setup jump target.
-  Value *addr32 = builder.CreateAnd(vcr, builder.getInt32(~0x3fu));
-  AddressExtender addressExtender(parent);
-  Value *jumpTarget = addressExtender.extend(addr32, builder.getInt32(HighAddrPc), builder.getPtrTy(), builder);
-
-  unsigned waveSize = m_pipelineState->getShaderWaveSize(m_shaderStage);
-  Value *execMask = builder.CreateIntrinsic(Intrinsic::amdgcn_ballot, builder.getIntNTy(waveSize), builder.getTrue());
-
-  Value *chainArgs[] = {jumpTarget, execMask, userDataVec, vgprArg, builder.getInt32(0)};
-#if LLVM_MAIN_REVISION && LLVM_MAIN_REVISION < 465197
-  // Old version of the code
-  Type *chainArgTys[] = {builder.getPtrTy(), builder.getIntNTy(waveSize), userDataVec->getType(), vgprArg->getType(),
-                         builder.getInt32Ty()};
-  FunctionType *funcTy = FunctionType::get(builder.getVoidTy(), chainArgTys, true);
-  auto func = Function::Create(funcTy, GlobalValue::ExternalLinkage, "llvm.amdgcn.cs.chain", parent->getParent());
-  builder.CreateCall(func, chainArgs);
-#else
-  // New version of the code (also handles unknown version, which we treat as
-  // latest)
-  Type *chainTys[] = {builder.getPtrTy(), builder.getIntNTy(waveSize), userDataVec->getType(), vgprArg->getType()};
-  builder.CreateIntrinsic(Intrinsic::amdgcn_cs_chain, chainTys, chainArgs);
-#endif
+  splitIntoI32(layout, builder, remainingArgs, vgprArgs);
+
+  // Fill exit information.
+  exitInfos.push_back(CpsExitInfo(jumpOp->getParent(), std::move(vgprArgs)));
+  // Branch to tailBlock.
+  auto *oldTerm = jumpOp->getParent()->getTerminator();
+  assert(isa<UnreachableInst>(oldTerm));
+  oldTerm->eraseFromParent();
+  builder.CreateBr(tailBlock);
```

Comment:
This is probably fine in practice, but a more robust solution for this sort of thing (without relying on the `unreachable` being there) is:

* Split the basic block after the `jumpOp`
* Insert the branch instruction

The tail of the old basic block remains as an unreachable basic block, to be cleaned up by a later pass. An unreachable basic block is dominated by everything, so even if there were other instructions in the tail that use some values, they would always be in SSA form.

Reply:
I am assuming the passes before should generate a valid form of cps.jump followed immediately by unreachable. The assertion will check against this. If there is any meaningful instruction happen after the cps.jump other than unreachable. Then I would assume there might be something wrong happened. So current code might help catch possible errors. Do you have any possible shape of valid IR but breaking the assumption here? I agree your suggestion could work in more cases. But it may be more useful to catch possible broken cases for now.

Reply:
No, I can't think of any plausible reason why we would end up with valid IR that doesn't fit the pattern. So yeah, I'm fine with that.